### PR TITLE
Prevent from complains in spdk_tgt when deleting a v2 volume

### PR DESCRIPTION
**Pull request by derekbit**
#### Which issue(s) this PR fixes:

Issue longhorn/longhorn#7568

#### What this PR does / why we need it:

spdk_tgt complains the improper sequence of deleting engines and replicas.
Replicas should be deleted before deleting its engine.

### CODE REVIEW

File: controller/setting_controller.go [hunk 1]
```diff
@@ -834,13 +834,13 @@ func (sc *SettingController) updateDataEngine(setting types.SettingName) error {
...
 
 func (sc *SettingController) cleanupInstanceManager(dataEngine longhorn.DataEngineType) error {
-	sc.logger.Infof("Cleaning up the instance manager for %v data engine", dataEngine)
 	imMap, err := sc.ds.ListInstanceManagersBySelectorRO("", "", longhorn.InstanceManagerTypeAllInOne, dataEngine)
 	if err != nil {
 		return errors.Wrapf(err, "failed to list instance managers for cleaning up %v data engine", dataEngine)
 	}
 
 	for _, im := range imMap {
+		sc.logger.Infof("Cleaning up the instance manager for %v data engine", dataEngine)
```
Comment (c3y1huang) [inline controller/setting_controller.go:843]:
```suggestion
		sc.logger.Infof("Cleaning up the instance manager %v for %v data engine", im.Name, dataEngine)
```

File: controller/volume_controller.go [hunk 2]
```diff
@@ -350,6 +350,28 @@ func (c *VolumeController) syncVolume(key string) (err error) {
 				}
 			}
 		}
+		if datastore.IsDataEngineV2(volume.Spec.DataEngine) {
+			// To prevent from the "no such device" error in spdk_tgt,
+			// remove the raid bdev before tearing down the replicas.
+			engines, err := c.ds.ListVolumeEnginesRO(volume.Name)
```
Comment (c3y1huang) [inline controller/volume_controller.go:356]: Should we consider the engine's DeletionTimestamp?

Comment (derekbit) [reply controller/volume_controller.go:356]: Although `engine.Spec.DeletionTimestamp` is set, it does not mean the engine (raid bdev) is already tear down. Hence, I only check the existence of engines here.
Do you think there is any side effect?

File: controller/setting_controller.go (hunk 1 above)
Comment (derekbit) [reply controller/setting_controller.go]: Updated. Thank you.

File: controller/volume_controller.go (hunk 2 above)
Comment (c3y1huang) [reply controller/volume_controller.go:356]: I see. No concerns.

**Review by c3y1huang — APPROVED** [review]

File: controller/volume_controller.go [hunk 3]
```diff
@@ -350,6 +350,28 @@ func (c *VolumeController) syncVolume(key string) (err error) {
...
+			if len(engines) > 0 {
+				replicaDeleted := false
+				for _, r := range replicas {
+					if r.DeletionTimestamp != nil {
+						replicaDeleted = true
+						break
+					}
+				}
+				if !replicaDeleted {
+					c.logger.Infof("Volume %v still has engines %v", volume.Name, engines)
```
Comment (innobead) [inline controller/volume_controller.go:369]:
```go
Infof("Volume (%s) %v still has engines %v, so skip deleting its replicas until the engines have been deleted", volume.Spec.DataEngine, volume.Name, engines)
```

**Review by innobead — CHANGES_REQUESTED** [review]
LGTM, just one comment.

File: controller/volume_controller.go (hunk 3 above)
Comment (derekbit) [reply controller/volume_controller.go]: Updated. Thank you.

**Review by innobead — APPROVED** [review]

**Pull request merged**